django/core/files/utils.py

```diff
@@ -1,16 +1,26 @@
 import os
+import pathlib
 
 from django.core.exceptions import SuspiciousFileOperation
 
 
-def validate_file_name(name):
-    if name != os.path.basename(name):
-        raise SuspiciousFileOperation("File name '%s' includes path elements" % name)
-
+def validate_file_name(name, allow_relative_path=False):
     # Remove potentially dangerous names
-    if name in {'', '.', '..'}:
+    if os.path.basename(name) in {'', '.', '..'}:
         raise SuspiciousFileOperation("Could not derive file name from '%s'" % name)
 
+    if allow_relative_path:
+        # Use PurePosixPath() because this branch is checked only in
+        # FileField.generate_filename() where all file paths are expected to be
+        # Unix style (with forward slashes).
+        path = pathlib.PurePosixPath(name)
+        if path.is_absolute() or '..' in path.parts:
```

Comment:
If `upload_to` returns an absolute path that already is in the MEDIA_ROOT it now throws an error saying detected path traversal attempt.  Is that desired behavior?

Reply:
Just hitting this exact issue now that 3.2.11 has dropped with this in it.

The problem is that calling `validate_file_name()` with `allow_relative_path=True` now disallows absolute paths. I'd expect allowing relative paths would be inclusive of absolute paths, so long as `..` isn't present (to prevent traversals).